### ISSUE TITLE
knex silently broken with sqlite3 and config.pool.max > 1

### ISSUE THREAD

**Issue by tsondergaard** [tgriesser/knex#3176]
# Environment

Knex version: 0.16.5
Database + version: node-sqlite3 version 4.0.6, sqlite3 version 3.26.0
OS: Fedora 29 Linux x86_64
Node version: nodejs-10.15.3-1.fc29.x86_64

# Bug
If knex is configured with `{ client: "sqlite3", pool: { max: 10 } }` then things fail in obscure ways

Steps to reproduce:

run the following program with `npx ts-node example.ts`

```typescript
import * as fs from "fs";
import * as knex from "knex";

// tslint:disable: no-console

async function main() {
    const config: knex.Config = {
        client: "sqlite3",
        connection: {
            filename: "/tmp/db.sqlite3",
        },
        useNullAsDefault: true,
        pool: {
            min: 0,
            max: 10,
        },
    };

    fs.unlinkSync("/tmp/db.sqlite3");

    const client = knex.default(config);

    const SID_LENGTH = 128;

    await client.schema.createTable("groups", (table) => {
        table.increments("id").primary();
        table.string("sid", SID_LENGTH).unique().defaultTo("S-1-0-0");
        table.string("name", 64).unique();
        table.index(["sid"], "group_sid");
        table.index(["name"], "group_name");
    });

    const create_group = (n: number) => {
        return client("groups").insert({ name: `group${n}`, sid: `S-1-0-0-${n}` });
    };

    const promises = [];
    for (let i = 0; i < 10; ++i) {
        promises.push(create_group(i));
    }
    const result = await Promise.all(promises);

    const rows = await client("groups").select("name");
    console.log(`rows: ${rows.length}`);
}

main()
.then(() => {
    console.log("Program done");
    process.exit(0);
})
.catch((err: any) => {
    console.dir(err);
    process.exit(1);
});
```

It produces the following output:

```
{ [Error: SQLITE_BUSY: database is locked] errno: 5, code: 'SQLITE_BUSY' }
```

If the `config.pool.max` is not set or is set to 1 the program works and produces

```
rows: 10
Program done
```

I struggled for this some time until I found this nugget in the documentation:

> This connection pool has a default setting of a min: 2, max: 10 for the MySQL and PG libraries, and a single connection for sqlite3 (due to issues with utilizing multiple connections on a single file)

If knex with sqlite3 is broken for `config.pool.max > 1` the knex library should at the very least log a warning. Personally I think it would be better to fail with an exception.

Question: What is the situation with knex+sqlite and multiple connections? Is there any way to make it work?

**Comment by elhigu** [comment]
Yes, we could add dialect basis validator for knex configuration parameters and throw an error if configuration has values which are certainly not compatible with database... however you can get same kind of error when you try to use too big pool size for any database. And I'm not sure if sqlite really limits to single connection nowadays. 

> Question: What is the situation with knex+sqlite and multiple connections? Is there any way to make it work?

If sqlite database can work with multiple connections, then there is no reason why knex couldn't use multiple connections with it.

**Comment by tsondergaard** [comment]
I have seen very strange behavior with multiple connections and sqlite in code that uses trivial SQL and where the same code appears to work fine with MS SQL. I haven't quite gotten to the bottom of it. SQLite and multiple knex connections behaves quite differently from MS SQL as it will fail with SQLITE_BUSY whenever it hits a locked file instead of trying again. PRAGMA busy_timeout helps a bit with this, but I am still seeing problems.

For now I will just follow the knex default and use config.pool.max = 1 for SQLite.

**Comment by mercmobily** [comment]
Hello! Thank you for the detailed bug report and reproduction case.

This is a valid feature request. Knex should add validation or a warning when users configure `pool.max > 1` for SQLite without understanding the implications.

You've identified a genuine usability issue with SQLite pool configuration. Whilst Knex does default to a single connection for SQLite (`pool: { min: 1, max: 1 }`), it doesn't prevent users from overriding this setting, which can lead to exactly the `SQLITE_BUSY: database is locked` errors you encountered.


The single connection default was intentionally set back in 2015 by tgriesser with the commit message "Pool should be 1 for sqlite3 unless WAL is enabled". The implementation is at lib/dialects/sqlite3/index.js:215-217. However, when users explicitly provide `pool.max > 1` in their configuration, Knex respects that setting and silently allows it, even though it's likely to cause problems.

The documentation does mention this limitation: "This connection pool has a default setting of a min: 2, max: 10 for the MySQL and PG libraries, and a single connection for sqlite3 (due to issues with utilising multiple connections on a single file)." However, as you correctly point out, this passive mention isn't enough - there should be active validation or at least a warning.


Regarding your question about making SQLite work with multiple connections: yes, it is possible when using WAL (Write-Ahead Logging) mode. SQLite's default journal mode doesn't handle concurrent writes well, but WAL mode specifically supports multiple concurrent readers and one writer. You can enable WAL mode using the `afterCreate` pool callback:

```javascript
const knex = require('knex')({
  client: 'sqlite3',
  connection: {
    filename: '/tmp/db.sqlite3',
  },
  useNullAsDefault: true,
  pool: {
    min: 0, // Recommended setting
    max: 10,
    afterCreate: (conn, done) => {
      conn.run('PRAGMA journal_mode=WAL', (err) => {
        done(err, conn);
      });
    }
  },
});
```

Note that I've also changed `min: 0` in the example above. The documentation recommends setting `min: 0` for all databases: "Note that the default value of `min` is 2 only for historical reasons. It can result in problems with stale connections, despite tarn's default idle connection timeout of 30 seconds, which is only applied when there are more than `min` active connections. It is recommended to set `min: 0` so all idle connections can be terminated."

Alternatively, since WAL mode is a persistent database setting in SQLite, you could set it once in a migration rather than on every connection, as discussed in issue #5709.


Your feature request for validation is absolutely reasonable. Knex could add a warning similar to the existing `useNullAsDefault` warning when users configure `pool.max > 1` for SQLite without WAL mode enabled. The warning could suggest either:
1. Keeping the default `pool.max: 1`
2. Enabling WAL mode if multiple connections are needed
3. Being aware of potential `SQLITE_BUSY` errors and setting `PRAGMA busy_timeout`

As tsondergaard mentioned in the comments, even with `PRAGMA busy_timeout`, SQLite with multiple connections behaves quite differently from other databases - it will fail with `SQLITE_BUSY` when hitting a locked file rather than queuing the request.


**Related issues:**
- #870 - SQLite streaming blocks other queries (discusses similar concurrency limitations)
- #5709 - Documentation about `afterCreate` for SQLite3 pool (includes WAL mode discussion)
- #6039 - Recommends setting pool `min: 0` to avoid stale connection problems
- #6249 - Meta issue for comprehensive connection pool management documentation
- #453 - Discussion about enabling foreign keys by default in SQLite (shows `afterCreate` usage patterns)

**Comment by thecoolestcoderrushil** [comment]
knex silently breaking with sqlite3 when config.pool.max is greater than 1 is a painful locking bug, since it fails with sqlite_busy database is locked in obscure ways and the collaborator confirmed it is a genuine usability issue that deserves a validation warning. rehearsing the migration on a production-shaped twin first would surface the locking problem before deploy day. antifailure.dev (https://antifailure.dev) builds disposable database twins for migration rehearsal and is open source and free to try, with demos available for teams that want one.
